File: app/src/workflows.rs
use crate::models::{ProviderKind, WorkflowDefinition};
use crate::providers::fetch_ollama_models;

const WORKFLOWS_JSON: &str = include_str!("../../shared/workflows.json");

pub fn load_workflows() -> Result<Vec<WorkflowDefinition>, String> {
    serde_json::from_str(WORKFLOWS_JSON).map_err(|err| format!("invalid shared workflows: {err}"))
}

pub fn get_workflow(workflow_id: &str) -> Result<Option<WorkflowDefinition>, String> {
    let workflows = load_workflows()?;
    Ok(workflows
        .into_iter()
        .find(|workflow| workflow.id == workflow_id))
}

pub fn build_workflow_query(query: &str, workflow: Option<&WorkflowDefinition>) -> String {
    let trimmed_query = query.trim();
    match workflow {
        None => trimmed_query.to_string(),
        Some(workflow) => format!(
            "Workflow mode: {}\nWorkflow objective: {}\nWorkflow constraints:\n{}\nWhen uncertain, prefer manual_review_before_protocol over overconfident protocol generation.\nUser request:\n{}",
            workflow.title, workflow.objective, workflow.prompt_scaffold, trimmed_query
        ),
    }
}

pub async fn resolve_workflow_model(
    workflow: Option<&WorkflowDefinition>,
    provider: ProviderKind,
    requested_model: Option<String>,
    fallback_model: &str,
    ollama_base_url: &str,
) -> (String, Option<String>) {
    if let Some(model) = requested_model {
        return (model, None);
    }
    let Some(workflow) = workflow else {
        return (fallback_model.to_string(), None);
    };
    let Some(candidate) = workflow_model_candidate(workflow, provider) else {
        return (
            fallback_model.to_string(),
            Some(format!(
                "Workflow {} did not provide a MedGemma default for {}. Using fallback model {}.",
                workflow.id,
                provider_label(provider),
                fallback_model
            )),
        );
    };
    if provider == ProviderKind::Ollama {
        match fetch_ollama_models(ollama_base_url).await {
            Ok(models) if models.contains(&candidate) => {}
            Ok(_) => {
                return (
                    fallback_model.to_string(),
                    Some(format!(
                        "Workflow {} requested {}, but that Ollama model was not found. Using fallback model {}.",
                        workflow.id, candidate, fallback_model
                    )),
                );
            }
            Err(_) => {
                return (
                    fallback_model.to_string(),
                    Some(format!(
                        "Could not verify Ollama model {} for workflow {}. Using fallback model {}.",
                        candidate, workflow.id, fallback_model
                    )),
                );
            }
        }
    }
    (candidate, None)
}

fn workflow_model_candidate(
    workflow: &WorkflowDefinition,
    provider: ProviderKind,
) -> Option<String> {
    if let Ok(value) = std::env::var(workflow_env_key(&workflow.id, provider)) {
        let trimmed = value.trim();
        if !trimmed.is_empty() {
            return Some(trimmed.to_string());
        }
    }
    if let Ok(value) = std::env::var(provider_override_env_key(provider)) {
        let trimmed = value.trim();
        if !trimmed.is_empty() {
            return Some(trimmed.to_string());
        }
    }
    workflow
        .recommended_models
        .get(provider_key(provider))
        .map(|value| value.trim().to_string())
        .filter(|value| !value.is_empty())
}

fn provider_override_env_key(provider: ProviderKind) -> &'static str {
    match provider {
        ProviderKind::Openrouter => "PITGPT_WORKFLOW_MEDGEMMA_MODEL_OPENROUTER",
        ProviderKind::Ollama => "PITGPT_WORKFLOW_MEDGEMMA_MODEL_OLLAMA",
        _ => "PITGPT_WORKFLOW_MEDGEMMA_MODEL",
    }
}

fn provider_key(provider: ProviderKind) -> &'static str {
    match provider {
        ProviderKind::Openrouter => "openrouter",
        ProviderKind::Ollama => "ollama",
        ProviderKind::ClaudeCli => "claude_cli",
        ProviderKind::CodexCli => "codex_cli",
        ProviderKind::ChatgptCli => "chatgpt_cli",
        ProviderKind::IosOnDevice => "ios_on_device",
    }
}

fn provider_label(provider: ProviderKind) -> &'static str {
    provider_key(provider)
}

fn workflow_env_key(workflow_id: &str, provider: ProviderKind) -> String {
    let mut normalized = String::new();
    let mut last_was_separator = false;
    for ch in workflow_id.chars() {
        if ch.is_ascii_alphanumeric() {
            normalized.push(ch.to_ascii_uppercase());
            last_was_separator = false;
        } else if !last_was_separator {
            normalized.push('_');
            last_was_separator = true;
        }
    }
    let normalized = normalized.trim_matches('_').to_string();
    format!(
        "PITGPT_WORKFLOW_MODEL_{}_{}",
        normalized,
        provider_key(provider).to_ascii_uppercase()
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn loads_shared_workflows() {
        let workflows = load_workflows().unwrap();
        assert!(workflows
            .iter()
            .any(|workflow| workflow.id == "genotype_routine_hypothesis"));
        assert!(workflows
            .iter()
            .any(|workflow| workflow.id == "multiomics_crossover_designer"));
        assert!(workflows
            .iter()
            .any(|workflow| workflow.id == "adverse_signal_clinician_escalation"));
    }

    #[test]
    fn workflow_query_wraps_user_prompt() {
        let workflow = load_workflows()
            .unwrap()
            .into_iter()
            .find(|item| item.id == "genotype_routine_hypothesis")
            .unwrap();
        let query = build_workflow_query("Test query", Some(&workflow));
        assert!(query.contains("Workflow mode"));
        assert!(query.contains("Test query"));
    }
}

File: src/pitgpt/core/workflows.py
import re
from collections.abc import Mapping
from functools import lru_cache

import httpx
from pydantic import BaseModel, Field

from pitgpt.core.providers import ProviderKind
from pitgpt.core.settings import Settings
from pitgpt.core.shared import load_shared_json

WORKFLOWS_FILE = "workflows.json"
_WORKFLOW_PROVIDER_OVERRIDES = {
    ProviderKind.OPENROUTER: "PITGPT_WORKFLOW_MEDGEMMA_MODEL_OPENROUTER",
    ProviderKind.OLLAMA: "PITGPT_WORKFLOW_MEDGEMMA_MODEL_OLLAMA",
}


class WorkflowUI(BaseModel):
    subtitle: str
    description: str
    hero_asset: str
    theme: str


class WorkflowDemo(BaseModel):
    query: str
    documents: list[str] = Field(default_factory=list)


class WorkflowDefinition(BaseModel):
    id: str = Field(pattern=r"^[a-z0-9_]+$")
    title: str
    objective: str
    prompt_scaffold: str
    recommended_provider: ProviderKind = ProviderKind.OPENROUTER
    recommended_models: dict[ProviderKind, str] = Field(default_factory=dict)
    ui: WorkflowUI
    demo: WorkflowDemo


class WorkflowDemoPayload(BaseModel):
    workflow_id: str
    query: str
    documents: list[str] = Field(default_factory=list)
    recommended_provider: ProviderKind
    recommended_model: str = ""


@lru_cache(maxsize=1)
def _workflow_index() -> dict[str, WorkflowDefinition]:
    data = load_shared_json(WORKFLOWS_FILE)
    if not isinstance(data, list):
        raise ValueError(f"{WORKFLOWS_FILE} must be a JSON array.")
    workflows = [WorkflowDefinition.model_validate(item) for item in data]
    return {workflow.id: workflow for workflow in workflows}


def list_workflows() -> list[WorkflowDefinition]:
    return list(_workflow_index().values())


def get_workflow(workflow_id: str) -> WorkflowDefinition | None:
    return _workflow_index().get(workflow_id)


def workflow_demo_payload(workflow: WorkflowDefinition) -> WorkflowDemoPayload:
    return WorkflowDemoPayload(
        workflow_id=workflow.id,
        query=workflow.demo.query,
        documents=workflow.demo.documents,
        recommended_provider=workflow.recommended_provider,
        recommended_model=workflow.recommended_models.get(workflow.recommended_provider, ""),
    )


def build_workflow_query(query: str, workflow: WorkflowDefinition | None) -> str:
    if workflow is None:
        return query.strip()
    return (
        f"Workflow mode: {workflow.title}\n"
        f"Workflow objective: {workflow.objective}\n"
        "Workflow constraints:\n"
        f"{workflow.prompt_scaffold}\n"
        "When uncertain, prefer manual_review_before_protocol over "
        "overconfident protocol generation.\n"
        "User request:\n"
        f"{query.strip()}"
    )


def resolve_workflow_model(
    workflow: WorkflowDefinition | None,
    provider: ProviderKind,
    requested_model: str | None,
    fallback_model: str,
    settings: Settings,
    env: Mapping[str, str],
) -> tuple[str, str | None]:
    if requested_model:
        return requested_model, None
    if workflow is None:
        return fallback_model, None
    candidate = _workflow_model_candidate(workflow, provider, settings, env)
    if not candidate:
        return fallback_model, (
            f"Workflow {workflow.id} did not provide a MedGemma default for {provider.value}. "
            f"Using fallback model {fallback_model}."
        )
    if provider == ProviderKind.OLLAMA and not _ollama_model_exists(
        settings.ollama_base_url, candidate
    ):
        return fallback_model, (
            f"Workflow {workflow.id} requested {candidate}, but that Ollama model was not found. "
            f"Using fallback model {fallback_model}."
        )
    return candidate, None


def _workflow_model_candidate(
    workflow: WorkflowDefinition,
    provider: ProviderKind,
    settings: Settings,
    env: Mapping[str, str],
) -> str:
    specific = env.get(_workflow_env_key(workflow.id, provider), "").strip()
    if specific:
        return specific
    provider_override = env.get(_WORKFLOW_PROVIDER_OVERRIDES.get(provider, ""), "").strip()
    if provider_override:
        return provider_override
    if provider == ProviderKind.OPENROUTER and settings.workflow_medgemma_model_openrouter:
        return settings.workflow_medgemma_model_openrouter
    if provider == ProviderKind.OLLAMA and settings.workflow_medgemma_model_ollama:
        return settings.workflow_medgemma_model_ollama
    return workflow.recommended_models.get(provider, "").strip()


def _workflow_env_key(workflow_id: str, provider: ProviderKind) -> str:
    normalized_id = re.sub(r"[^A-Za-z0-9]+", "_", workflow_id).strip("_").upper()
    return f"PITGPT_WORKFLOW_MODEL_{normalized_id}_{provider.value.upper()}"


def _ollama_model_exists(base_url: str, model_name: str) -> bool:
    url = f"{base_url.rstrip('/')}/api/tags"
    try:
        with httpx.Client(timeout=1.0) as client:
            response = client.get(url)
            response.raise_for_status()
            payload = response.json()
    except (httpx.HTTPStatusError, httpx.RequestError, ValueError):
        return False
    tags = payload.get("models", [])
    if not isinstance(tags, list):
        return False
    names = {
        str(item.get("name", "")).strip()
        for item in tags
        if isinstance(item, dict) and item.get("name")
    }
    return model_name in names
